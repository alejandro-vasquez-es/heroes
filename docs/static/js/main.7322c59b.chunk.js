(this["webpackJsonpheroes-app"]=this["webpackJsonpheroes-app"]||[]).push([[0],{19:function(e,r,a){},34:function(e,r,a){"use strict";a.r(r);var t=a(1),c=a(12),s=a.n(c),n=a(5),i=Object(t.createContext)(),l=a(6),o="[auth] Login",h="[auth] Logout",j=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},r=arguments.length>1?arguments[1]:void 0;switch(r.type){case o:return Object(l.a)(Object(l.a)({},r.payload),{},{logged:!0});case h:return{logged:!1};default:return e}},u=a(3),d=a(2),b=a(0),p=function(e){var r=e.children,a=Object(t.useContext)(i).user,c=Object(d.f)(),s=c.pathname,n=c.search;return localStorage.setItem("lastPath",s+n),a.logged?r:Object(b.jsx)(d.a,{to:"/login"})},m=function(e){var r=e.children;return Object(t.useContext)(i).user.logged?Object(b.jsx)(d.a,{to:"/"}):r},O=function(){var e=Object(d.g)(),r=Object(t.useContext)(i).dispatch;return Object(b.jsxs)("div",{className:"container mt-5",children:[Object(b.jsx)("h1",{children:"Login page"}),Object(b.jsx)("hr",{}),Object(b.jsx)("button",{className:"btn btn-outline-primary",onClick:function(){r({type:o,payload:{name:"User Name"}});var a=localStorage.getItem("lastPath");e(a,{replace:!0})},children:"Login"})]})},v=function(){var e=Object(t.useContext)(i),r=e.user,a=e.dispatch,c=r.name,s=Object(d.g)();return Object(b.jsxs)("nav",{className:"px-4 pl-4 navbar navbar-expand-sm navbar-dark bg-dark",children:[Object(b.jsx)(u.b,{className:"navbar-brand",to:"/",children:"Asociaciones"}),Object(b.jsx)("div",{className:"navbar-collapse",children:Object(b.jsxs)("div",{className:"navbar-nav",children:[Object(b.jsx)(u.c,{className:function(e){return"nav-item nav-link "+(e.isActive?"active":"")},to:"/marvel",children:"Marvel"}),Object(b.jsx)(u.c,{className:function(e){return"nav-item nav-link "+(e.isActive?"active":"")},to:"/dc",children:"DC"}),Object(b.jsx)(u.c,{className:function(e){return"nav-item nav-link "+(e.isActive?"active":"")},to:"/search",children:"Search"})]})}),Object(b.jsx)("div",{className:"order-3 navbar-collapse collapse w-100 dual-collapse2 d-flex justify-content-end",children:Object(b.jsxs)("ul",{className:"ml-auto navbar-nav",children:[Object(b.jsx)("span",{className:"nav-item nav-link text-info",children:c}),Object(b.jsx)("button",{onClick:function(){a({type:h}),s("/login",{replace:!0})},id:"logout",className:"nav-item nav-link btn",children:"Logout"})]})})]})},x=(a(19),function(e){var r=e.id,a=e.superhero,t=e.publisher,c=e.alter_ego,s=e.first_appearance,n=e.characters,i=r.split("-")[0];return Object(b.jsxs)(u.b,{to:"/".concat(i,"/hero/").concat(r),className:"my-card",children:[Object(b.jsx)("img",{src:"./assets/heroes/".concat(r,".jpg"),className:"img img-responsive",alt:a}),Object(b.jsx)("div",{className:"profile-name",children:a}),Object(b.jsx)("div",{className:"profile-position",children:c}),Object(b.jsx)("div",{className:"profile-overview",children:Object(b.jsx)("div",{className:"profile-overview",children:Object(b.jsx)("div",{className:"row",children:Object(b.jsxs)("div",{className:"col-ms-4",children:[Object(b.jsx)("h3",{children:t}),Object(b.jsxs)("p",{children:["Primera aparici\xf3n: ",Object(b.jsx)("br",{}),s]}),c!==n&&Object(b.jsx)("p",{children:n})]})})})})]})}),f=[{id:"dc-batman",superhero:"Batman",publisher:"DC Comics",alter_ego:"Bruce Wayne",first_appearance:"Detective Comics #27",characters:"Bruce Wayne"},{id:"dc-superman",superhero:"Superman",publisher:"DC Comics",alter_ego:"Kal-El",first_appearance:"Action Comics #1",characters:"Kal-El"},{id:"dc-flash",superhero:"Flash",publisher:"DC Comics",alter_ego:"Jay Garrick",first_appearance:"Flash Comics #1",characters:"Jay Garrick, Barry Allen, Wally West, Bart Allen"},{id:"dc-green",superhero:"Green Lantern",publisher:"DC Comics",alter_ego:"Alan Scott",first_appearance:"All-American Comics #16",characters:"Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz"},{id:"dc-arrow",superhero:"Green Arrow",publisher:"DC Comics",alter_ego:"Oliver Queen",first_appearance:"More Fun Comics #73",characters:"Oliver Queen"},{id:"dc-wonder",superhero:"Wonder Woman",publisher:"DC Comics",alter_ego:"Princess Diana",first_appearance:"All Star Comics #8",characters:"Princess Diana"},{id:"dc-martian",superhero:"Martian Manhunter",publisher:"DC Comics",alter_ego:"J'onn J'onzz",first_appearance:"Detective Comics #225",characters:"Martian Manhunter"},{id:"dc-robin",superhero:"Robin/Nightwing",publisher:"DC Comics",alter_ego:"Dick Grayson",first_appearance:"Detective Comics #38",characters:"Dick Grayson"},{id:"dc-blue",superhero:"Blue Beetle",publisher:"DC Comics",alter_ego:"Dan Garret",first_appearance:"Mystery Men Comics #1",characters:"Dan Garret, Ted Kord, Jaime Reyes"},{id:"dc-black",superhero:"Black Canary",publisher:"DC Comics",alter_ego:"Dinah Drake",first_appearance:"Flash Comics #86",characters:"Dinah Drake, Dinah Lance"},{id:"marvel-spider",superhero:"Spider Man",publisher:"Marvel Comics",alter_ego:"Peter Parker",first_appearance:"Amazing Fantasy #15",characters:"Peter Parker"},{id:"marvel-captain",superhero:"Captain America",publisher:"Marvel Comics",alter_ego:"Steve Rogers",first_appearance:"Captain America Comics #1",characters:"Steve Rogers"},{id:"marvel-iron",superhero:"Iron Man",publisher:"Marvel Comics",alter_ego:"Tony Stark",first_appearance:"Tales of Suspense #39",characters:"Tony Stark"},{id:"marvel-thor",superhero:"Thor",publisher:"Marvel Comics",alter_ego:"Thor Odinson",first_appearance:"Journey into Myster #83",characters:"Thor Odinson"},{id:"marvel-hulk",superhero:"Hulk",publisher:"Marvel Comics",alter_ego:"Bruce Banner",first_appearance:"The Incredible Hulk #1",characters:"Bruce Banner"},{id:"marvel-wolverine",superhero:"Wolverine",publisher:"Marvel Comics",alter_ego:"James Howlett",first_appearance:"The Incredible Hulk #180",characters:"James Howlett"},{id:"marvel-daredevil",superhero:"Daredevil",publisher:"Marvel Comics",alter_ego:"Matthew Michael Murdock",first_appearance:"Daredevil #1",characters:"Matthew Michael Murdock"},{id:"marvel-hawkeye",superhero:"Hawkeye",publisher:"Marvel Comics",alter_ego:"Clinton Francis Barton",first_appearance:"Tales of Suspense #57",characters:"Clinton Francis Barton"},{id:"marvel-cyclops",superhero:"Cyclops",publisher:"Marvel Comics",alter_ego:"Scott Summers",first_appearance:"X-Men #1",characters:"Scott Summers"},{id:"marvel-silver",superhero:"Silver Surfer",publisher:"Marvel Comics",alter_ego:"Norrin Radd",first_appearance:"The Fantastic Four #48",characters:"Norrin Radd"}],g=function(e){var r=e.publisher,a=void 0===r?"marvel":r,c=Object(t.useMemo)((function(){return function(e){var r=arguments.length>1&&void 0!==arguments[1]?arguments[1]:f;if(!["DC Comics","Marvel Comics"].includes(e))throw new Error("".concat(e," is not a valid publisher"));return r.filter((function(r){return r.publisher===e}))}(a)}),[a]);return Object(b.jsx)("div",{className:"mt-4 row rows-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn",children:c.map((function(e){return Object(b.jsx)(x,Object(l.a)({},e),e.id)}))})},C=function(){return Object(b.jsxs)("div",{children:[Object(b.jsx)("h1",{children:"Dc page"}),Object(b.jsx)(g,{publisher:"DC Comics"})]})},_=function(){return Object(b.jsxs)("div",{children:[Object(b.jsx)("h1",{children:"Marvel page"}),Object(b.jsx)(g,{publisher:"Marvel Comics"})]})},N=a(13),y=a.n(N),S=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:"";return""===e?[]:(e=e.toLowerCase(),f.filter((function(r){return r.superhero.toLowerCase().includes(e)})))},k=a(7),M=function(){var e=Object(d.g)(),r=Object(d.f)(),a=y.a.parse(r.search).q,c=void 0===a?"":a,s=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},r=Object(t.useState)(e),a=Object(n.a)(r,2),c=a[0],s=a[1];return[c,function(e){var r=e.target;s(Object(l.a)(Object(l.a)({},c),{},Object(k.a)({},r.name,r.value)))},function(){s(e)}]}({searchText:c}),i=Object(n.a)(s,2),o=i[0],h=i[1],j=o.searchText,u=Object(t.useState)(S(j)),p=Object(n.a)(u,2),m=p[0],O=p[1];Object(t.useEffect)((function(){O(S(j))}),[j]),Object(t.useEffect)((function(){e("?q=".concat(j))}),[j]);var v=function(r){r.preventDefault(),e("?q=".concat(j))};return Object(b.jsxs)("div",{children:[Object(b.jsx)("h1",{children:"Search page"}),Object(b.jsx)("hr",{}),Object(b.jsxs)("div",{className:"row",children:[Object(b.jsxs)("div",{className:"col-5",children:[Object(b.jsx)("h4",{children:"Buscar"}),Object(b.jsx)("hr",{}),Object(b.jsxs)("form",{onSubmit:v,children:[Object(b.jsx)("input",{type:"text",placeholder:"Search for a hero",className:"form-control",name:"searchText",autoComplete:"off",value:j,onChange:h}),Object(b.jsx)("button",{type:"submit",className:"btn brn-outline mt-1",onClick:v,children:"Search..."})]})]}),Object(b.jsxs)("div",{className:"col-7",children:[Object(b.jsx)("h4",{children:"Results"}),Object(b.jsx)("hr",{}),""===c?Object(b.jsx)("div",{className:"alert-info alert animate__fadeIn animate__animated",children:"Search for a superhero."}):0===m.length&&Object(b.jsxs)("div",{id:"error",className:"alert alert-danger animate__fadeIn animate__animated",children:['No results for "',c,'"']}),m.map((function(e){return Object(b.jsx)(x,Object(l.a)({},e),e.id)}))]})]})]})},D=function(){var e=Object(d.g)(),r=Object(d.h)(),a=r.heroId,c=r.pub,s=Object(t.useMemo)((function(){return function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:"";return f.find((function(r){return r.id===e}))}(a)}),[a]);if(!s)return Object(b.jsx)(d.a,{to:"/"});var n="/assets/heroes/".concat(s.id,".jpg");return Object(b.jsxs)("div",{className:"mt-5 row",children:[Object(b.jsx)("div",{className:"col-3",children:Object(b.jsx)("img",{src:n,alt:s.superhero,className:"img-thumbnail animate__animated animate__fadeInLeft"})}),Object(b.jsxs)("div",{className:"col-9 d-flex flex-column",children:[Object(b.jsx)("h1",{children:s.superhero}),Object(b.jsxs)("ul",{className:"list-group list-group-flush",children:[Object(b.jsxs)("li",{className:"list-group-item",children:[Object(b.jsx)("b",{children:"Alter ego: "}),s.alter_ego]}),Object(b.jsxs)("li",{className:"list-group-item",children:[Object(b.jsx)("b",{children:"Publisher: "}),s.publisher]}),Object(b.jsxs)("li",{className:"list-group-item",children:[Object(b.jsx)("b",{children:"First Appearance: "}),s.first_appearance]})]}),Object(b.jsx)("h5",{className:"mt-4",children:"Characters"}),Object(b.jsx)("p",{children:s.characters}),Object(b.jsx)("button",{className:"btn btn-outline-info align-self-end",onClick:function(){e("/".concat(c))},children:"Regresar"})]})]})},w=function(){return Object(b.jsxs)(b.Fragment,{children:[Object(b.jsx)(v,{}),Object(b.jsx)("div",{className:"container mt-3",children:Object(b.jsxs)(d.d,{children:[Object(b.jsx)(d.b,{path:"marvel",element:Object(b.jsx)(_,{})}),Object(b.jsx)(d.b,{path:"dc",element:Object(b.jsx)(C,{})}),Object(b.jsx)(d.b,{path:":pub/hero/:heroId",element:Object(b.jsx)(D,{})}),Object(b.jsx)(d.b,{path:"search",element:Object(b.jsx)(M,{})}),Object(b.jsx)(d.b,{path:"/",element:Object(b.jsx)(_,{})})]})})]})},A=function(){return Object(b.jsx)(u.a,{children:Object(b.jsxs)(d.d,{children:[Object(b.jsx)(d.b,{path:"/login",element:Object(b.jsx)(m,{children:Object(b.jsx)(O,{})})}),Object(b.jsx)(d.b,{path:"/*",element:Object(b.jsx)(p,{children:Object(b.jsx)(w,{})})})]})})},B=function(){return JSON.parse(localStorage.getItem("user"))||{logged:!1}},J=function(){var e=Object(t.useReducer)(j,{},B),r=Object(n.a)(e,2),a=r[0],c=r[1];return Object(t.useEffect)((function(){a&&localStorage.setItem("user",JSON.stringify(a))}),[a]),Object(b.jsx)(i.Provider,{value:{user:a,dispatch:c},children:Object(b.jsx)(A,{})})};s.a.render(Object(b.jsx)(J,{}),document.getElementById("root"))}},[[34,1,2]]]);
//# sourceMappingURL=main.7322c59b.chunk.js.map