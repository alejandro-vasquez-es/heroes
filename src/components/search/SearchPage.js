import React, { useEffect, useState } from 'react'
import { useNavigate, useLocation } from 'react-router-dom';

import queryString from 'query-string';
import { getHeroesByName } from '../../helpers/getHeroesByName.js';
import { useForm } from '../../hooks/useForm.js'
import { HeroCard } from '../hero/HeroCard.js';

export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search)

    const [formValues, handleInputChange] = useForm({
        searchText: q

    });

    const { searchText } = formValues;
    // const heroesFiltered = getHeroesByName('');
    const [heroesFiltered, setHeroesFiltered] = useState(getHeroesByName(searchText))

    useEffect(() => {
        setHeroesFiltered(getHeroesByName(searchText))
    }, [searchText])

    useEffect(() => {
        navigate(`?q=${searchText}`)
    }, [searchText, navigate])

    const handleSearch = (e) => {

        e.preventDefault();
        navigate(`?q=${searchText}`);

    }


    return (
        <div>
            <h1>Search page</h1>
            <hr />

            <div className='row'>

                <div className='col-5'>
                    <h4>Buscar</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Search for a hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={handleInputChange}
                        />

                        <button
                            type='submit'
                            className='btn brn-outline mt-1'
                            onClick={handleSearch}
                        >
                            Search...
                        </button>

                    </form>
                </div>

                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />

                    {
                        (q === '') ?
                            <div className="alert-info alert animate__fadeIn animate__animated">Search for a superhero.</div>
                            : (heroesFiltered.length === 0)
                            && <div id='error' className='alert alert-danger animate__fadeIn animate__animated'>No results for "{q}"</div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }

                </div>

            </div>

        </div>
    )
}
