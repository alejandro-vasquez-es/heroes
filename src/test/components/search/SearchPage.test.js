import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../components/search/SearchPage.js";


const mockNavigate = jest.fn();

jest.mock('react-router-dom', ()=> ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('<SearchPage /> tests', () => {
  
    test('should display propertly with default values', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search for a superhero.');

    });
    
    test('should display batman card and the input with the queryString', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )
      
        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper ).toMatchSnapshot();
        
    });
    
    test('should display an error if do not find a hero', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                    <SearchPage />
            </MemoryRouter>
        )

        expect( wrapper.find('#error').exists() ).toBeTruthy();
        expect( wrapper ).toMatchSnapshot();

        
    });
    
    test('should call navigate to the new url', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=superman']} >
                <SearchPage />
            </MemoryRouter>
        )
        
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        })
        
        wrapper.find('form').simulate('submit', {
            preventDefault: ()=> {}
        })

        expect(mockNavigate ).toHaveBeenCalledWith('?q=superman');
        
    });
    
    
});
