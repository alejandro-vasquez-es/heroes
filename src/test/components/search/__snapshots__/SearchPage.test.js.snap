// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage /> tests should display an error if do not find a hero 1`] = `
<div>
  <h1>
    Search page
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search for a hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn brn-outline mt-1"
          onClick={[Function]}
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__fadeIn animate__animated"
        id="error"
      >
        No results for "
        batman123
        "
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchPage /> tests should display batman card and the input with the queryString 1`] = `
<div>
  <h1>
    Search page
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search for a hero"
          type="text"
          value="batman"
        />
        <button
          className="btn brn-outline mt-1"
          onClick={[Function]}
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <a
        className="my-card"
        href="/dc/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
